### PR TITLE
SDL-0289 Support for Set Language Separately 

### LINKED ISSUE #1313: [SDL 0289] Support for Set Language Separately
Proposal: [Support for Set Language Separately](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0289-set-language-separately.md)

> This proposal provides support for updating to two separate languages.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/963

### PR DESCRIPTION
**Pull request by zhouxin627**
Fixes #1313 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

### Summary
A new SdlManagerListener method would need to be added, and the old one should be deprecated.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by zhouxin627** [comment]
@theresalech Hi, this PR is now ready for review.

**Comment by theresalech** [comment]
@zhouxin627 thank you!  This feature is not planned for the upcoming Java Suite 4.11 release (scheduled for 2020-04-29), but may be able to be included in the following release if prioritized by the Steering Committee.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java [hunk 1]
```diff
@@ -62,8 +62,9 @@
 	 * is not supported.
 	 *
 	 * @param language The language of the connected head unit the manager is trying to update the configuration.
+	 * @param hmiLanguage The hmiLanguage of the connected head unit the manager is trying to update the configuration.
 	 * @return An object of LifecycleConfigurationUpdate if the head unit language is supported,
 	 * otherwise null to indicate that the language is not supported.
 	 */
-	LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language);
+	LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language, Language hmiLanguage);
```
Comment (JulianKast) [inline android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java:69]: Per the proposal, The old method has to be deprecated.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java [hunk 2]
```diff
@@ -999,10 +1001,19 @@ public Builder setMinimumRPCVersion(final Version minimumRPCVersion) {
 		}
 
 		/**
-		 * Sets the Language of the App
+		 * Sets the VR+TTS Language of the App
+		 * @param language the desired language to be used on the VR+TTS of the connected module
+		 */
+		public Builder setLanguage(final Language language){
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java]: OK. This has been reverted.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java [hunk 3]
```diff
@@ -999,10 +1001,19 @@ public Builder setMinimumRPCVersion(final Version minimumRPCVersion) {
...
+			sdlManager.language = language;
+			return this;
+		}
+
+		/**
+		 * Sets the display/HMI Language of the App
 		 * @param hmiLanguage the desired language to be used on the display/HMI of the connected module
 		 */
-		public Builder setLanguage(final Language hmiLanguage){
+		public Builder setHMILanguage(final Language hmiLanguage){
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java]: OK. This has been reverted.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java [hunk 4]
```diff
@@ -62,8 +62,9 @@
 	 * is not supported.
 	 *
 	 * @param language The language of the connected head unit the manager is trying to update the configuration.
+	 * @param hmiLanguage The hmiLanguage of the connected head unit the manager is trying to update the configuration.
```
[the remark this answers is not in the document]
Comment (zhouxin627) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java]: We have changed the description of the parameters

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java (hunk 1 above)
Comment (zhouxin627) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManagerListener.java:84]: This has been fixed.

**Comment by theresalech** [comment]
@zhouxin627 can you please confirm if this PR is ready for re-review?

**Comment by zhouxin627** [comment]
@theresalech Yes, this is ready for re-review.

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java [hunk 5]
```diff
@@ -199,19 +199,44 @@ public void onError(String info, Exception e) {
...
 						case FR_CA:
+							isNeedUpdate = true;
 							appName = APP_NAME_FR;
 							break;
 						default:
-							return null;
+							break;
+					}
+					if (isNeedUpdate) {
+						return new LifecycleConfigurationUpdate(appName,null,TTSChunkFactory.createSimpleTTSChunks(ttsName), null);
```
Comment (JulianKast) [inline android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java:235]: Add spaces after commas here, like 
```(appName, null, TTSChunkFactory.createSimpleTTSChunks(ttsName), null);```

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java [hunk 6]
```diff
@@ -199,19 +199,44 @@ public void onError(String info, Exception e) {
 
 				@Override
 				public LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language){
-					String appName;
+					return null;
+				}
+
+				@Override
+				public LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language, Language hmiLanguage){
```
Comment (JulianKast) [inline android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java:206]: Add space before curly brace.
```
managerShouldUpdateLifecycle(Language language, Language hmiLanguage) {
```

File: android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java [hunk 7]
```diff
@@ -199,19 +199,44 @@ public void onError(String info, Exception e) {
 
 				@Override
 				public LifecycleConfigurationUpdate managerShouldUpdateLifecycle(Language language){
```
Comment (JulianKast) [inline android/hello_sdl_android/src/main/java/com/sdl/hellosdlandroid/SdlService.java:201]: Just to keep up with good formatting to keep all code consistent could you please add a space before the curly brace?

```
managerShouldUpdateLifecycle(Language language) {
```

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java [hunk 8]
```diff
@@ -675,7 +677,7 @@ public void onTransportEvent(List<TransportRecord> connectedTransports, boolean
 					}
 				}
 
-				proxy = new SdlProxyBase(proxyBridge, context, appName, shortAppName, isMediaApp, hmiLanguage,
+				proxy = new SdlProxyBase(proxyBridge, context, appName, shortAppName, isMediaApp, language,
```
Comment (JulianKast) [inline android/sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java:680]: Same thing here if they are using the old method

**Comment by JulianKast** [comment]
@zhouxin627 is this ready for re-review?

**Comment by zhouxin627** [comment]
@JulianKast Hi, this is ready for re-review.

**Comment by JulianKast** [comment]
Changes also need to be made for the javaSE project. javaSE has their own classes that are similar to android. The corresponding sdlService.java class is in the hello_sdl_java folder, the rest of the corresponding classes are in the javaSE folder. If you open the hello_sdl_java project in IntelliJ you should be able to modify and easily apply your changes to the javaSE project.

**Comment by zhouxin627** [comment]
Hi @JulianKast,  Thank you very much for your prompt. It is really helpful. But in fact, one of the problems we face is that no one in our team has the JavaSE development or test experience. It is difficult to complete the modification in a short period of time. So could you please change it instead of us, or could you please give us another option if there is. We will be very grateful.

**Comment by JulianKast** [comment]
> Hi @JulianKast, Thank you very much for your prompt. It is really helpful. But in fact, one of the problems we face is that no one in our team has the JavaSE development or test experience. It is difficult to complete the modification in a short period of time. So could you please change it instead of us, or could you please give us another option if there is. We will be very grateful.

The changes will be the same as you did with android, to the same-named classes and methods that are in different folders, if you apply the changes, I will do the testing.

**Comment by zhouxin627** [comment]
@JulianKast Thank you for your help. We added the javaSE code in 9fe10b7. It is ready for review.

**Comment by zhouxin627** [comment]
@JulianKast Hi, this is ready for re-review.

**Comment by JulianKast** [comment]
@zhouxin627 PR is merged!

**Pull request merged**